src/main/cpp/src/datetime_parser.cu

```diff
@@ -0,0 +1,668 @@
+/*
+ * Copyright (c) 2023-2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "datetime_parser.hpp"
+
+#include <iostream>
+#include <vector>
+
+#include <cuda/std/cassert>
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/null_mask.cuh>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/detail/valid_if.cuh>
+
+#include <cudf/lists/list_device_view.cuh>
+#include <cudf/lists/lists_column_device_view.cuh>
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/strings/string_view.cuh>
+
+#include <cudf/reduction.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/utilities/span.hpp>
+#include <cudf/wrappers/timestamps.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/binary_search.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/logical.h>
+#include <thrust/optional.h>
+#include <thrust/transform.h>
+#include <thrust/tuple.h>
+
+using column                   = cudf::column;
+using column_device_view       = cudf::column_device_view;
+using column_view              = cudf::column_view;
+using lists_column_device_view = cudf::detail::lists_column_device_view;
+using size_type                = cudf::size_type;
+using string_view              = cudf::string_view;
+using struct_view              = cudf::struct_view;
+using table_view               = cudf::table_view;
+
+namespace {
+
+/**
+ * Represents local date time in a time zone.
+ */
+struct timestamp_components {
+  int32_t year;  // max 6 digits
```

Comment:
If we support 6 digit years can we make it clear in the javadocs?

Reply:
Done.